program/src/duplicate_block_proof.rs

```diff
@@ -3,14 +3,80 @@ use {
     crate::{
         error::SlashingError,
         shred::{Shred, ShredType},
+        sigverify::SignatureVerification,
         state::{ProofType, SlashingProofData},
     },
     bytemuck::try_from_bytes,
-    solana_program::{clock::Slot, msg, pubkey::Pubkey},
+    solana_program::{
+        account_info::{next_account_info, AccountInfo},
+        clock::Slot,
+        hash::Hash,
+        msg,
+        pubkey::Pubkey,
+    },
+    solana_signature::SIGNATURE_BYTES,
     spl_pod::primitives::PodU32,
+    std::slice::Iter,
 };
 
+/// The verification instruction occurs immediately before the slashing
+/// instruction
+const SIGVERIFY_INSTRUCTION_RELATIVE_INDEX: i64 = -1;
+/// Both shreds are verified in the same instruction
+const NUM_VERIFICATIONS_IN_INSTRUCTION: usize = 2;
+
+#[repr(C)]
+#[derive(Clone, Copy, PartialEq, Eq)]
+/// Signature verification context required for duplicate block
+/// proof verification
+pub struct DuplicateBlockProofContext<'a> {
+    pub(crate) expected_pubkey: &'a Pubkey,
+    pub(crate) expected_shred1_merkle_root: &'a Hash,
+    pub(crate) expected_shred2_merkle_root: &'a Hash,
+    pub(crate) expected_shred1_signature: &'a [u8; SIGNATURE_BYTES],
+    pub(crate) expected_shred2_signature: &'a [u8; SIGNATURE_BYTES],
```

Comment:
tiny nit: can this be laid out in the same order as the instruction? That way you could eventually just cast the instruction data to this type and avoid doing any work to create it, ie
```suggestion
    pub(crate) expected_shred1_merkle_root: &'a Hash,
    pub(crate) expected_shred1_signature: &'a [u8; SIGNATURE_BYTES],
    pub(crate) expected_shred2_merkle_root: &'a Hash,
    pub(crate) expected_shred2_signature: &'a [u8; SIGNATURE_BYTES],
```

Reply:
yep done here b5b5b0fb4fa76cbb0b6036dd0c15bb994994dc61

> just cast the instruction data to this type and avoid doing any work to create it

For this are you suggesting that we don't actually copy the data over when doing the instruction introspection but instead just cast the instruction data?

Reply:
Yeah that's the idea, but that would require breaking a lot of the nice things that you've done for parsing instruction data. That can be for the far future if ever needed